Mathlib/SetTheory/ZFC/Basic.lean

```diff
@@ -509,11 +510,108 @@ end PSet
 def ZFSet : Type (u + 1) :=
   Quotient PSet.setoid.{u}
 
+namespace ZFSet
+
+/-- Turns a pre-set into a ZFC set. -/
+def mk : PSet → ZFSet :=
+  Quotient.mk''
+
+@[simp]
+theorem mk_eq (x : PSet) : @Eq ZFSet ⟦x⟧ (mk x) :=
+  rfl
+
+@[simp]
+theorem mk_out : ∀ x : ZFSet, mk x.out = x :=
+  Quotient.out_eq
+
+/-- A set function is "definable" if it is the image of some n-ary `PSet`
+  function. This isn't exactly definability, but is useful as a sufficient
+  condition for functions that have a computable image. -/
+class Definable (n) (f : (Fin n → ZFSet.{u}) → ZFSet.{u}) where
+  /-- Turns a definable function into an n-ary `PSet` function. -/
+  out : (Fin n → PSet.{u}) → PSet.{u}
+  /-- A set function `f` is the image of `Definable.out f`. -/
+  mk_out xs : mk (out xs) = f (mk <| xs ·) := by simp
+
+attribute [simp] Definable.mk_out
+
+/-- An abbrev of `ZFSet.Definable` for unary functions. -/
+abbrev Definable₁ (f : ZFSet.{u} → ZFSet.{u}) := Definable 1 (fun s ↦ f (s 0))
+
+/-- A simpler constructor for `ZFSet.Definable₁`. -/
+abbrev Definable₁.mk {f : ZFSet.{u} → ZFSet.{u}}
+    (out : PSet.{u} → PSet.{u}) (mk_out : ∀ x, ⟦out x⟧ = f ⟦x⟧) :
+    Definable₁ f where
+  out xs := out (xs 0)
+  mk_out xs := mk_out (xs 0)
+
+/-- Turns a unary definable function into a unary `PSet` function. -/
+abbrev Definable₁.out (f : ZFSet.{u} → ZFSet.{u}) [Definable₁ f] :
+    PSet.{u} → PSet.{u} :=
+  fun x ↦ Definable.out (fun s ↦ f (s 0)) ![x]
+
+lemma Definable₁.mk_out {f : ZFSet.{u} → ZFSet.{u}} [Definable₁ f]
+    {x : PSet} :
+    .mk (out f x) = f (.mk x) :=
+  Definable.mk_out ![x]
+
+/-- An abbrev of `ZFSet.Definable` for binary functions. -/
+abbrev Definable₂ (f : ZFSet.{u} → ZFSet.{u} → ZFSet.{u}) := Definable 2 (fun s ↦ f (s 0) (s 1))
+
```

Comment:
```suggestion
```

Reply:
Oh I'm sorry, mobile line-wrapping had me think there were two spaces here instead of one.